MM-35721: Clicking an Avatar should open Profile Popover

sinansonmez opened the pull request:
#### Summary
Clicking an avatar in thread item will open profile popover on
- the root post in the center channel
- the global threads page

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/19348

#### Screenshots
**before**

https://user-images.githubusercontent.com/37421564/153732631-8c993b1a-6c05-44eb-814d-d6838511752e.mov

**after**

https://user-images.githubusercontent.com/37421564/153732637-76a4ebfa-3716-40e7-ad8c-4ff0416e78a2.mov

#### Release Note
```release-note
NONE
```

mattermod commented:
Hello @sinansonmez,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -20,6 +20,9 @@ import Avatar from 'components/widgets/users/avatar';
 
 import './avatars.scss';
 import {getMissingProfilesByIds} from 'mattermost-redux/actions/users';
```
Done. Also good to know for future PRs

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -50,27 +57,47 @@ const displayNameGetter = makeDisplayNameGetter();
 
 function UserAvatar({
     userId,
-    overlayProps,
```
Yo are right, I missed it. Done

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -50,27 +57,47 @@ const displayNameGetter = makeDisplayNameGetter();
     ...props
 }: {
     userId: UserProfile['id'];
     overlayProps: Partial<ComponentProps<typeof SimpleTooltip>>;
```
Done

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -50,27 +57,47 @@ const displayNameGetter = makeDisplayNameGetter();
     const user = useSelector((state: GlobalState) => selectUser(state, userId)) as UserProfile | undefined;
-    const name = useSelector((state: GlobalState) => displayNameGetter(state, true)(user));
+
+    const overlay = React.createRef<MMOverlayTrigger>();
```
I am passsing this to OverlayTrigger as ref

M-ZubairAhmed replied:
I just checked it just to be sure, without ref functionality is not effected.

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -50,27 +57,47 @@ const displayNameGetter = makeDisplayNameGetter();
+
+    const overlay = React.createRef<MMOverlayTrigger>();
+
+    const getProfilePictureURL = (): string => {
+        if (userId) {
+            return Utils.imageURLForUser(userId);
+        }
+        return '';
+    };
```
Good point. Done

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -50,27 +57,47 @@ const displayNameGetter = makeDisplayNameGetter();
+            >
+                <Avatar
+                    url={imageURLForUser(userId, user?.last_picture_update)}
+                    tabIndex={0}
+                    size={props.size}
```
I wasn't passing `{...props}` because it was complaining due to unexpected `%s` element in props. Actually overlayProps was creating the issue. Since we removed it above, there is no problem anymore. 

In the `Avatar` element, I didn't see `style` prop and in the old version there was no style prop passed, so I assume it is fine

sinansonmez commented on e2e/cypress/integration/collapsed_reply_threads/replies_spec.js:
```diff
@@ -119,6 +125,39 @@ describe('Collapsed Reply Threads', () => {
         cy.get('article.ThreadItem').find('.activity').should('have.text', '1 new reply');
     });
 
+    it('should open popover when avatar is clicked', () => {
```
👍

sinansonmez commented on e2e/cypress/integration/collapsed_reply_threads/replies_spec.js:
```diff
@@ -119,6 +125,39 @@ describe('Collapsed Reply Threads', () => {
+
+        // # Get thread footer of last post and find avatars
+        cy.uiGetPostThreadFooter(postForAvatar.id).find('.Avatars').find('button').each(($el) => {
+            // # Click the avatar
```
Done

sinansonmez replied:
I deleted the empty line before line 139, because lint check was failing due to this

Willyfrog reviewed (changes requested):
Looking good, minus some comments :)

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -17,9 +18,11 @@ import {imageURLForUser} from 'utils/utils';
 
 import './avatars.scss';
-import {getMissingProfilesByIds} from 'mattermost-redux/actions/users';
+import * as Utils from 'utils/utils';
```
Line 25 deleted

sinansonmez commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -50,27 +57,39 @@ const displayNameGetter = makeDisplayNameGetter();
-    const name = useSelector((state: GlobalState) => displayNameGetter(state, true)(user));
+
+    const overlay = React.createRef<MMOverlayTrigger>();
+    const profilePictureURL = userId ? Utils.imageURLForUser(userId) : '';
```
Yeah you are both right. Changes are done accordingly

M-ZubairAhmed reviewed (approved):
All the changes looks good to me. Thank you for working on them

M-ZubairAhmed commented:
Adding @furqanmlk as this PR also has an E2E test

M-ZubairAhmed commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/150985

furqanmlk reviewed (approved):
Working as expected

![image](https://user-images.githubusercontent.com/6909890/154144463-6b0e943b-638d-4cff-80a8-b65898ff91f4.png)

jgilliam17 reviewed (commented):
Thanks @sinansonmez for your contribution
Noticed few issues

1. Cypress test case is failing, See [report ](https://automation-dashboard.vercel.app/cycles/cdb79c29-f2b4-4303-99d6-388471a0ce2b)
I also created a Zephyr test case and added the test id. See suggestion.
2.  Tooltip with user name is missing; tooltip is present when hovering over + icon. Also avatars should be slightly overlapping etc. 

See what we currently have on community.
<img width="188" alt="Screen Shot 2022-02-15 at 9 26 42 PM" src="https://user-images.githubusercontent.com/52937121/154184903-01e1a4f8-b17f-4477-b282-b7d05d5df5fb.png">
on test server 
<img width="197" alt="Screen Shot 2022-02-15 at 10 01 48 PM" src="https://user-images.githubusercontent.com/52937121/154188065-d52e60a3-4cd3-4258-94e2-96dfcab03935.png">

3.  On global threads, clicking on the avatar also opens that thread in the thread view.  @matthewbirtch should we suppress thread open on avatar click? 
We are currently not suppressing thread open on avatar click on global threads, but do so on main channel view thread avatars
<img width="1315" alt="Screen Shot 2022-02-15 at 9 25 04 PM" src="https://user-images.githubusercontent.com/52937121/154185312-fc523841-a6a7-4962-b1e5-686d041ac450.png">

sinansonmez commented:
Thanks for the feedback @jgilliam17, please see my comments below,

> 1. Cypress test case is failing, See [report ](https://automation-dashboard.vercel.app/cycles/cdb79c29-f2b4-4303-99d6-388471a0ce2b)
>    I also created a Zephyr test case and added the test id. See suggestion.

The tests were passing in my local but I will check the report to fix them

> 2\. Tooltip with user name is missing; tooltip is present when hovering over + icon. Also avatars should be slightly overlapping etc.

For tooltip, 
- I kept the tooltip if avatars are grouped together in single avatar 
![Screen Shot 2022-02-16 at 14 24 22](https://user-images.githubusercontent.com/37421564/154273693-fe30ee5e-43b3-4a2d-9ddf-cb1b9e7be832.png)

- I removed it intentionally for single avatar, to keep the same behaviour as message. In the screenshot below, when you hover purple profile image, you don't see tooltip for username but you click to open a profile popover. Please let me know if you would like to keep both username tooltip on hover and profile popover for single avatar
I will fix overlapping issue. 
![Screen Shot 2022-02-16 at 14 03 09](https://user-images.githubusercontent.com/37421564/154270423-a99afbd8-3cc8-4d4f-90d9-e3852de16097.png)

> 3\. On global threads, clicking on the avatar also opens that thread in the thread view.  @matthewbirtch should we suppress thread open on avatar click?
>     We are currently not suppressing thread open on avatar click on global threads, but do so on main channel view thread avatars

On global threads, the page opens with empty thread view (red part in the screenshot) to show all replies. Since the thread view is already there, showing profile popover and opening the thread could be useful with a single click on avatar

![Screen Shot 2022-02-16 at 14 16 51](https://user-images.githubusercontent.com/37421564/154272579-6c0609f4-583b-4d34-a955-66977a29ccc4.png)

Please let me know about final decision @matthewbirtch

matthewbirtch commented:
> * I removed it intentionally for single avatar, to keep the same behaviour as message. In the screenshot below, when you hover purple profile image, you don't see tooltip for username but you click to open a profile popover. Please let me know if you would like to keep both username tooltip on hover and profile popover for single avatar

Please keep the tooltip when hovering over individual avatars in the thread footer


> We are currently not suppressing thread open on avatar click on global threads, but do so on main channel view thread avatars
> 
> On global threads, the page opens with empty thread view (red part in the screenshot) to show all replies. Since the thread view is already there, showing profile popover and opening the thread could be useful with a single click on avatar

> Please let me know about final decision @matthewbirtch

I think it's fair to assume that if a user clicks an avatar that they aren't intending to open the thread, so I would suppress opening the thread in this case and only open the profile popover

sinansonmez commented:
/e2e-test

mattermod commented:
You don't have permissions to trigger this command.
 It's only available for organization members.

mattermod commented:
Triggering e2e testing with options:
```<nil>```

jgilliam17 reviewed (approved):
Thanks @sinansonmez for a quick fix.
Tested, looks good to merge.
- Verified thread footer is functioning as before. New option to click on participant's avatar opens user profile, profile dismisses on `esc`. 
- Verified click on thread participant's avatar is not opening thread view.
- [E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/152247-41c3367-onprem-ent-webapp-pr-9804-mattermost/mm-ee-test41c3367/mochawesome.html)looks good, no PR related failures. New test passes.
<img width="752" alt="Screen Shot 2022-02-17 at 4 41 07 PM" src="https://user-images.githubusercontent.com/52937121/154575467-9fdab1f7-8035-47bf-abc0-5a33bbff491e.png">

sinansonmez commented:
Thanks to you and other reviewers for the support @Willyfrog
